// viber-bot/test/viber-client-test.js
"use strict";

const ViberClient = require(__dirname + '/../lib/viber-client');
const TestEnvironmentConfiguration = require(__dirname + "/util/test-environment-configuration");
const ViberBot = require(__dirname + "/../lib/viber-bot");
const express = require('express');

function startServer(port, requestHandler, onFinish) {
	const app = express();
	app.post('/*', requestHandler);
	const server = app.listen(port, () => {
		return onFinish(null, server);
	});
}

module.exports = {
	testSendRequestHandlesOnError: test => {
		test.expect(2);
		const bot = new ViberBot(TestEnvironmentConfiguration.MockLogger, {
			authToken: "123AB",
			name: "Test",
			avatar: "http://avatar.com/image.jpg"
		});

		function requestHandler(req, res) {
			return res.status(500).send({ error: 'something blew up' });
		}

		return startServer(0, requestHandler, function(err, server) {
			test.ok(!err);

			const apiUrl = "http://127.0.0.1:" + server.address().port;
			const client = new ViberClient(TestEnvironmentConfiguration.MockLogger, bot, apiUrl, []);

			client.getAccountInfo().then(function(data) {
				test.ok(false, "did not receive an error");
				server.close();
				test.done();
			}, function(err) {
				test.ok(err);
				server.close();
				test.done();
			});
		});
	}
};
